### PR TITLE
VarSet Auto-Complete Tuning

### PR DESCRIPTION
**Pull request by bmurphy1976**
## Description

This PR implements a new strategy for auto completing VarSet references. It replaces fiddly string manipulation logic with an implementation based on and re-using the existing ExpressionTokenizer.  This makes autocomplete more consistent and predicable in practice.  It subjectively feels more like using auto complete (or intellisense) in a modern code editor and less like whatever FeeCAD was doing previously.

## Details

This has bothered me considerably since I started learning FreeCAD and I've wanted to find ways to contribute for awhile now, thus this PR is my attempt to clean this up.

The most obvious improvement is that when using `<<VarSet>>.` syntax, the auto complete will pop up immediately after typing `.`.  Without this change you have to type at least one character (`<<VarSet>>.A`) to trigger the popup, which is really frustrating when you don't remember the property name.

## AI Usage

- [X] This PR is not unverified AI output, I take responsibility for it, and all communication from my side in this PR is done by me personally.

I used Codex 5.5 to help design the implementation and learn code base.  I used Fable to review and polish the implementation while ensuring I understood the intimate details, the implementation is minimal, sound, and conforms to project norms.

Assisted-by Codex 5.5
Assisted-by Anthropic Fable

## Before and After

Here's a short video showing pop-up funkiness before the change:

https://github.com/user-attachments/assets/b8c8f0b6-fdb7-4ac8-af72-d6d1545475d7

Here's a short video showing responsiveness after the change:

https://github.com/user-attachments/assets/1ab499cf-fdd6-4f8d-90e5-ee711c64dd41

## Implementation Details

Alas there is no associated ticket or discussion.  This is a personal frustration of mine that I attacked head on.  It does affect the GUI, but the implementation is self contained and I could not find any evidence of my changes creating any compatibility issues or affecting third party plugins.  I have inlined some commentary on the PR to help document my decision making process for anybody who is interested.

### CODE REVIEW

File: src/App/ExpressionTokenizer.cpp [hunk 1]
```diff
@@ -34,7 +34,7 @@ using namespace App;
 // Code below inspired by blog entry:
 // https://john.nachtimwald.com/2009/07/04/qcompleter-and-comma-separated-tags/
 
-QString ExpressionTokenizer::perform(const QString& prefix, int pos)
+QString ExpressionTokenizer::extractCompletionPrefix(const QString& prefix, int pos)
```
Comment (bmurphy1976) [inline src/App/ExpressionTokenizer.cpp:37]: I changed the name of this method to be consistent with the newly introduced splitCompletionPath function as well as make it more explanatory. Perform was probably OK when this class contained one method, but with a new method it now doesn't do a good job conveying what the method does.

Because this is an interface change, I spent a lot of time digging for other references and couldn't find anything concerning.  The method is only used in a few key places inside the code base addressed by this PR and I don't think it's exposed in a way that would allow external plugins to access it.

Thus I think the chance of this introducing a backwards compatibility issue is low.

File: src/App/ExpressionTokenizer.cpp [hunk 2]
```diff
@@ -155,3 +155,69 @@ QString ExpressionTokenizer::perform(const QString& prefix, int pos)
...
+{
+bool isCompletionPathComponent(int token)
+{
+    return token == ExpressionParser::IDENTIFIER || token == ExpressionParser::INTEGER
+        || token == ExpressionParser::STRING || token == ExpressionParser::UNIT
+        || token == ExpressionParser::ONE || token == ExpressionParser::CELLADDRESS;
+}
+}  // namespace
+
+QStringList ExpressionTokenizer::splitCompletionPath(const QString& path) const
```
Comment (bmurphy1976) [inline src/App/ExpressionTokenizer.cpp:169]: This is the heart of the new implementation.  Instead of using a lot of stringy search/find/index logic we build on top of the existing tokenizer and use a similar tokenization strategy to break the text input more reliably.

File: src/App/ExpressionTokenizer.h [hunk 3]
```diff
@@ -33,7 +34,18 @@ namespace App
 class AppExport ExpressionTokenizer
 {
 public:
-    QString perform(const QString& text, int pos);
```
Comment (bmurphy1976) [inline src/App/ExpressionTokenizer.h:36]: As mentioned above, we rename this and add a comment to help future devs understand more easily what the method does.

File: src/Gui/ExpressionCompleter.cpp [hunk 4]
```diff
@@ -784,104 +782,16 @@ QString ExpressionCompleter::pathFromIndex(const QModelIndex& index) const
...
 {
-    QStringList resultList;
-    std::string path = input.toUtf8().constData();
-    if (path.empty()) {
-        return resultList;
+    if (input.isEmpty()) {
+        return {};
     }
 
-    int retry = 0;
```
Comment (bmurphy1976) [inline src/Gui/ExpressionCompleter.cpp:819]: This is the old string parsing implementation.  Because it's more of a UI wrapper, we get two benefits here:

1. We use the new tokenizer method to improve usability
2. We get "business logic" out of what is really a UI concern and improve layering/code structure slightly

File: tests/src/App/Expression.cpp [hunk 5]
```diff
@@ -48,22 +48,76 @@ class Expression: public ::testing::Test
...
+              QStringList({QStringLiteral("Constraints"), QString()}));
+
+    // No trailing separator: single component, no empty tail.
+    EXPECT_EQ(tokenizer.splitCompletionPath(QStringLiteral("Sketch")),
+              QStringList({QStringLiteral("Sketch")}));
+    // Trailing '#' offers the object completion level.
+    EXPECT_EQ(tokenizer.splitCompletionPath(QStringLiteral("Doc#")),
+              QStringList({QStringLiteral("Doc"), QString()}));
+    // Quoted document label followed by an object.
+    EXPECT_EQ(tokenizer.splitCompletionPath(QStringLiteral("<<My Doc>>#Obj.")),
```
Comment (bmurphy1976) [inline tests/src/App/Expression.cpp:95]: Basic tests, nothing really exotic here.  I tried to pay particular attention to less simple edge cases such as the use of << or >>, and #.

File: src/App/ExpressionTokenizer.cpp [hunk 6]
```diff
@@ -155,3 +155,69 @@ QString ExpressionTokenizer::perform(const QString& prefix, int pos)
 
     return completionPrefix;
 }
+
+namespace
+{
```
Comment (aw-engineer) [inline src/App/ExpressionTokenizer.cpp:160]: It looks like isCompletionPathComponent is called one place (ExpressionTokenizer).  Does this unnamed namespace offer a benefit over a static (or regular) class method?

Comment (bmurphy1976) [reply src/App/ExpressionTokenizer.cpp:160]: My C++ expertise is a couple decades out of date here so forgive me if I don't quite get this exactly correct.

I didn't particularly care for this convention myself, but it's used elsewhere in the codebase (search for requestTargetsDocument and isSuppressedQtWarning) and I had it out with the LLMs trying to understand WHY.  Per my explorations with Claude where I'm trying to fully understand this pattern that I'm not intimately familiar with:

>It's effectively the same as a file-static function.  A static class method is the one that would actually be different: it has to be declared in ExpressionTokenizer.h, which puts a single-file implementation detail into the public header of an exported class (and recompiles everything that includes it).  The unnamed namespace keeps it out of the header entirely.

The convention itself comes from the C++ Core Guidelines:

https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#rs-unnamed2

which there's evidence of desired compliance via .clang-tidy.

That said, I return to my original statement:  I was writing C++ in the 90s, not the 2020s.  I'm happy to change this PR if needed.

Comment (aw-engineer) [reply src/App/ExpressionTokenizer.cpp:160]: Thanks, this agrees with my understanding after reviewing https://en.cppreference.com/cpp/language/namespace.

**Comment by aw-engineer** [comment]
I cherry-picked the changeset onto my development branch to test the software change.  I exercised a few use cases I have run into using VarSets recently

# Software Versions

## Baseline

```
commit 6c19d73badacf374795a4d16eb37cacfe587451b (origin2/additivepipe-binder)
Author: Andrew Wilson <87784967+aw-engineer@users.noreply.github.com>
Date:   Sat Aug 1 18:23:35 2026 +0000

    Prefer ProfileBased::getTopoShapeVerifiedFace instead of Pipe::getSectionShape
    
    Adds unit test for SubShapeBinder profiles with FeaturePipe
    Updates FeaturePipe Task dialog

commit 5d05aaaf32c670134acc8d74c2db3f1589be56ce (main)
Author: wwmayer <wmayer@freecad.org>
Date:   Sat Aug 1 15:36:00 2026 +0200

    Part: Add further unit tests
    
    This is to verify that the add() method of Compound, Shell and CompSolid
    still works.
    
    See https://github.com/FreeCAD/FreeCAD/issues/31666
```

## Candidate

```
commit 6741779bc65f42b80d2ccfc328c5d63eaab0f65c (HEAD -> additivepipe-binder)
Author: Bryan Murphy <bmurphy@mediafly.com>
Date:   Fri Jul 3 10:38:49 2026 -0500

    Tuned autocomplete behavior using the tokenizer for more robust
    parsing.

commit 6c19d73badacf374795a4d16eb37cacfe587451b (origin2/additivepipe-binder)
Author: Andrew Wilson <87784967+aw-engineer@users.noreply.github.com>
Date:   Sat Aug 1 18:23:35 2026 +0000

    Prefer ProfileBased::getTopoShapeVerifiedFace instead of Pipe::getSectionShape
    
    Adds unit test for SubShapeBinder profiles with FeaturePipe
    Updates FeaturePipe Task dialog

commit 5d05aaaf32c670134acc8d74c2db3f1589be56ce (main)
Author: wwmayer <wmayer@freecad.org>
Date:   Sat Aug 1 15:36:00 2026 +0200

    Part: Add further unit tests
    
    This is to verify that the add() method of Compound, Shell and CompSolid
    still works.
    
    See https://github.com/FreeCAD/FreeCAD/issues/31666
```

# Tests

## Baseline

**VarSet Label**

| testcase | description | expected | observed | result |
|----------|-|-|-|-|
| L-01     | {varset label} {dot} | dropdown of variable name candidates | 'failed to parse expression' | fail
| L-02     | {varset label} {select fqn autocomplete} | user input replaced with selected candidate | user input *prepended* with selected candidate | fail
| L-03     | {varset label, angle brackets} {dot} | dropdown of variable name candidates | 'failed to parse expression' | fail
| L-04     | {varset label, angle brackets} {dot} {subobject} {dot} | dropdown of variable name candidates | 'failed to parse expression' | fail

**VarSet Internal Name**

| testcase | description                  | expected | observed | result |
|-|------------------------------|-|-|-|
| IN-01 | {varset internal name} {dot} | dropdown of variable name candidates | dropdown of variable name candidates | pass
| IN-02 | {varset internal name} {dot} {partial known variable name} | dropdown of variable name candidates, including the correct one | dropdown of variable name candidates, including the correct one | pass

## Candidate

As described in the PR, the change addresses **L-03** test case above.

| testcase | description | expected | observed | result |
|----------|-|-|-|-|
| L-03     | {varset label, angle brackets} {dot} | dropdown of variable name candidates | dropdown of variable name candidates | pass

This provides behavior consistent with using Internal Names in the expression editor.

File: src/App/ExpressionTokenizer.cpp (hunk 6 above)
Comment (aw-engineer) [reply src/App/ExpressionTokenizer.cpp:160]: Referencing the [FreeCAD - Code Review Guide](https://freecad.github.io/DevelopersHandbook/bestpractices/codereview)

> 15. (C++ only) Code should not be put in anonymous namespace.

I do find that external visibility of methods (in c and c++) aids in their unit testing.

File: tests/src/App/Expression.cpp [hunk 7]
```diff
@@ -48,22 +48,76 @@ class Expression: public ::testing::Test
...
+    EXPECT_EQ(tokenizer.splitCompletionPath(QStringLiteral("<<A.B>>.")),
+              QStringList({QStringLiteral("<<A.B>>"), QString()}));
+    // Lone '.' offers the root completion level (current-object property access).
+    EXPECT_EQ(tokenizer.splitCompletionPath(QStringLiteral(".")),
+              QStringList({QString()}));
+    // Number tokens are valid component fragments and concatenate across whitespace,
+    // matching extractCompletionPrefix() ("My Cube 1" -> "MyCube1").
+    EXPECT_EQ(tokenizer.splitCompletionPath(QStringLiteral("My Cube 1.")),
+              QStringList({QStringLiteral("MyCube1"), QString()}));
+    // "A1" lexes as a cell address, not an identifier, and must still be a component.
```
Comment (chennes) [inline tests/src/App/Expression.cpp]: This isn't true, A1 lexes as an identifier:

Image

(261 is an IDENTIFIER,  CELLADDRESS is 266). If you want it to lex as a CELLADDRESS you need dollar signs.

Comment (bmurphy1976) [reply tests/src/App/Expression.cpp]: I think I understand.  Does https://github.com/FreeCAD/FreeCAD/pull/31489/changes/4b4e74e0c619b2a1f2bbd13157b36f6d79217a0e work?

**Review by chennes — APPROVED** [review]

**Comment by chennes** [comment]
@pieterhijma do you want to look at this before we merge it? It worked well in my testing, and the code is good.

**Comment by maxwxyz** [comment]
I think this is superseded by https://github.com/FreeCAD/FreeCAD/pull/30531

**Comment by chennes** [comment]
@maxwxyz they certainly *conflict*, but they are also clearly independent solutions to the problem, and are quite different from one another. At a quick glance it's not obvious to me that one is superior to the other. I haven't looked at #30531 in detail, but this PR has the clear advantage in size :).

**Comment by bmurphy1976** [comment]
@chennes @maxwxyz there's likely some value in exploring a combination of these two approaches.  My approach using the tokenizer is more focused, replacing messy string parsing with a proper tokenization strategy and nothing else.  I don't know how this strategy performs with large models or property lists, but it doesn't use any techniques that I would anticipate introduces a performance regression over baseline.  If anything it probably performs better.

But I definitely do not do any property ranking or caching.  I'm a little skeptical of the caching.  It makes sense as a quick fix for the performance issue, but I don't know if that's also a problem with baseline or introduced by the fuzzy matching returning or doing too much.  That needs some investigation on my end.

I also don't know how much value the fuzzy matching actually improves UX.  I'm not advanced enough of a user to know.

What I can do is try to make a 3rd PR that combines the best of the two and we can then collectively decide which of the 3 approaches makes the most sense or get there in stages.
